libqrexec/exec.c

```diff
@@ -105,4 +139,240 @@ void do_fork_exec(const char *cmdline, int *pid, int *stdin_fd, int *stdout_fd,
+                (use_bare_path ? service_length : command->service_descriptor_length);
+            if (sizeof service_full_path <= total_path_length) {
+                fputs("qrexec internal error: miscalculated the buffer size needed", stderr);
+                abort();
+            }
+            memcpy(service_full_path, qubes_rpc_directories[i].string, directory_length);
+            memcpy(service_full_path + directory_length, command->command, total_path_length - directory_length);
+            service_full_path[total_path_length] = '\0';
+
+            if (!qubes_connect(s, service_full_path, total_path_length)) {
+                *stdout_fd = *stdin_fd = s;
+                if (stderr_fd) {
+                    *stderr_fd = -1;
+                }
+                *pid = -1;
+                set_nonblock(s);
+                buffer_append(stdin_buffer, command->service_descriptor, strlen(command->service_descriptor) + 1);
+                return 0;
+            }
+            switch (errno) {
+            // These cannot happen
+            case EFAULT:       // all of our parameters are in valid memory
+            case EINVAL:       // we passed valid parameters
+            case EBADF:        // this is a valid file descriptor
+            case ENOTSOCK:     // `s` was created by a call to `socket()`
+            case ENETUNREACH:  // cannot happen for AF_UNIX
+            case EADDRNOTAVAIL:// cannot happen for AF_UNIX
+            case EINPROGRESS:  // this socket is blocking
+            case EAGAIN:       // ditto
+#if EAGAIN != EWOULDBLOCK
+            case EWOULDBLOCK:  // ditto
+#endif
+            case EINTR:        // we already check for this and retry
+            case EAFNOSUPPORT: // the kernel supports AF_UNIX
+            case EALREADY:     // we have not connected this yet!
+            case EPROTO:       // no remote peer to mess up the protocol
+            default:
+                perror("unexpected return from connect()");
+                abort();
+            // Filesystem misconfigured
+            case ENOENT:
+            case ENOTDIR:
+                // These errors would also happen with `execve()`
+                break;
+            // These should not happen
+            case EPROTOTYPE:
+            case ETIMEDOUT:
+            case EPERM:
+            case EIO:
+            case EISDIR:
+                // Socket server problem and/or misconfiguration.  Fail the whole connection.
+                // (We do not want to fall back because the user may have
+                // overriden behavior for security reasons, and because
+                // fail-fast is much easier to debug).
+                goto out;
+            // Found a file instead of a socket
+            case EACCES:
+            case ECONNREFUSED:
+                close(s);
+                s = -1;
+                return do_fork_exec(command->username, command->command, pid, stdin_fd, stdout_fd, stderr_fd);
```

Comment:
You need to re-add the `QUBESRPC` string at some point, otherwise `exec_qubes_rpc_if_requested` won't work (or change the logic in `do_exec`).